torchtune/datasets/multimodal/_llava_instruct.py

```diff
@@ -0,0 +1,113 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+from typing import Any, Dict, Optional, Union
+
+from torchtune.data import ShareGPTToMessages
+from torchtune.datasets._packed import PackedDataset
+from torchtune.datasets._sft import SFTDataset
+from torchtune.modules.transforms import Transform
+
+
+def llava_instruct_dataset(
+    model_transform: Transform,
+    *,
+    coco_image_dir: str,
+    source: str = "liuhaotian/LLaVA-Instruct-150K",
+    column_map: Optional[Dict[str, str]] = None,
+    new_system_prompt: Optional[str] = None,
+    image_tag: Optional[str] = "<image>",
+    train_on_input: bool = True,
+    packed: bool = False,
+    split: str = "train",
+    data_files: str = "llava_instruct_150k.json",
+    **load_dataset_kwargs: Dict[str, Any],
+) -> Union[SFTDataset, PackedDataset]:
+    """
+    Support for family of image + text datasets similar to
+    `LLaVA-Instruct-150K <https://huggingface.co/datasets/liuhaotian/LLaVA-Instruct-150K>_`
+    from Hugging Face Datasets.
+
+    To use this dataset, you must first download the COCO Train 2017 image dataset. You can do so
+    by visiting https://cocodataset.org/#download or downloading it directly:
+
+    .. code-block:: bash
+
+        wget -c http://images.cocodataset.org/zips/train2017.zip
+        unzip train2017.zip -d coco/
+
+    Then, you must pass in the directory containing all the images to the ``image_dir`` argument
+    so each file can be loaded as PIL images in the transform. In the example above, you would
+    set ``image_dir="coco/train2017"``.
+
+    Args:
+        model_transform (Transform): model-specific transform that takes in a sample dict and applies custom
+            transforms on the keys. The tokenizer used by the model should be encapsulated in the model transform
+            and should operate on the "messages" field. Any model-specific image transforms should operate on
+            the "images" field. The keys returned by the model should be aligned with the
+            expected inputs into the model.
```

Comment:
This is very general and probably the most important part of setting this up correctly. I would try to give an explicit example here or something

Reply:
once the flamingo transform lands, it would be easy to point to this

Reply:
Can we add a TODO or a follow-up then? I wanna make sure it's very clear how to use these builders, rn I feel like it's not